File: Omeotte/Omeotte/Stats.m
//
//  Stats.m
//  Omeotte
//
//  Created by Jovito Royeca on 8/23/13.
//  Copyright (c) 2013 JJJ Software. All rights reserved.
//

#import "Stats.h"

@implementation Stats

@synthesize tower;
@synthesize wall;

@synthesize bricks;
@synthesize gems;
@synthesize recruits;

@synthesize quarries;
@synthesize magics;
@synthesize dungeons;

+(void) applyEffect:(Effect)effect onCurrent:(Stats*)current  orOpponent:(Stats*) opponent
{
    switch (effect->target)
    {
        case Current:
        {
            [current setStatField:effect->field withValue:effect->value];
            break;
        }
        case Opponent:
        {
            [opponent setStatField:effect->field withValue:effect->value];
            break;
        }
    }
}

+(void) evaluate:(Eval)eval onCurrent:(Stats*) current  orOpponent:(Stats*) opponent
{
    
}

-(int) statField:(StatField)field
{
    switch (field)
    {
        case Tower:
        {
            return self.tower;
        }
        case Wall:
        {
            return self.wall;
        }
        case Bricks:
        {
            return self.bricks;
        }
        case Gems:
        {
            return self.gems;
        }
        case Recruits:
        {
            return self.recruits;
        }
        case Quarries:
        {
            return self.quarries;
        }
        case Magics:
        {
            return self.magics;
        }
        case Dungeons:
        {
            return self.dungeons;
        }
        case None:
        default:
        {
            return 0;
        }
    }
}

-(void) setStatField:(StatField)field withValue:(int)value
{
    switch (field)
    {
        case Tower:
        {
            self.tower += value;
            break;
        }
        case Wall:
        {
            self.wall += value;
            
            if (self.wall < 0)
            {
                self.tower += self.wall;
                self.wall = 0;
            }
            break;
        }
        case Bricks:
        {
            self.bricks += value;
            
            if (self.bricks < 0)
            {
                self.bricks = 0;
            }
            break;
        }
        case Gems:
        {
            self.gems += value;
            
            if (self.gems < 0)
            {
                self.gems = 0;
            }
            break;
        }
        case Recruits:
        {
            self.recruits += value;
            
            if (self.recruits < 0)
            {
                self.recruits = 0;
            }
            break;
        }
        case Quarries:
        {
            self.quarries += value;
            
            if (self.quarries < 0)
            {
                self.quarries = 0;
            }
            break;
        }
        case Magics:
        {
            self.magics += value;
            
            if (self.magics < 0)
            {
                self.magics = 0;
            }
            break;
        }
        case Dungeons:
        {
            self.dungeons += value;
            
            if (self.dungeons < 0)
            {
                self.dungeons = 0;
            }
            break;
        }
        case None:
        default:
        {
            break;
        }
    }
}



-(void) setStats:(Stats*)src
{
    [self setStatField:Tower withValue:src.tower];
    [self setStatField:Wall withValue:src.wall];
    [self setStatField:Bricks withValue:src.bricks];
    [self setStatField:Gems withValue:src.gems];
    [self setStatField:Recruits withValue:src.recruits];
    [self setStatField:Quarries withValue:src.quarries];
    [self setStatField:Magics withValue:src.magics];
    [self setStatField:Dungeons withValue:src.dungeons];
}

-(int) compare:(int)op1 with:(int)op2
{
    int result = 0;
    
    if (op1 < op2)
    {
        result = -1;
    }
    else if (op1 == op2)
    {
        result = 0;
    }
    else if (op1 > op2)
    {
        result = 1;
    }
    
    return result;
}

@end

File: Omeotte/Omeotte/stats.c
//
//  Omeotte.c
//  Omeotte
//
//  Created by Jovito Royeca on 9/18/13.
//  Copyright (c) 2013 JJJ Software. All rights reserved.
//

#include <stdlib.h>

#include "Omeotte.h"

OStats createStats()
{
    OStats stats = (OStats) malloc(sizeof(OStats));

    if (stats)
    {
        stats->tower = 0;
        stats->wall = 0;
        stats->bricks = 0;
        stats->gems = 0;
        stats->recruits = 0;
        stats->quarries = 0;
        stats->magics = 0;
        stats->dungeons = 0;
    
        return stats;
    }

    return 0;
}

OEffect createEffect()
{
    OEffect effect = (OEffect) malloc(sizeof(OEffect));
    
    if (effect)
    {
        effect->field = None;
        effect->value = 0;
        effect->target = Current;
    
        return effect;
    }
    
    return 0;
}

void applyEffect(OStats current, OStats opponent, OEffect effect)
{
    switch (effect->target)
    {
        case Current:
        {
            setStatsField(current, effect->field, effect->value);
            break;
        }
        case Opponent:
        {
            setStatsField(opponent, effect->field, effect->value);
            break;
        }
    }
}

void evaluate(OStats current, OStats opponent, OEval eval)
{
    
}

int statsField(OStats stats, OStatsField field)
{
    switch (field)
    {
        case Tower:
        {
            return stats->tower;
        }
        case Wall:
        {
            return stats->wall;
        }
        case Bricks:
        {
            return stats->bricks;
        }
        case Gems:
        {
            return stats->gems;
        }
        case Recruits:
        {
            return stats->recruits;
        }
        case Quarries:
        {
            return stats->quarries;
        }
        case Magics:
        {
            return stats->magics;
        }
        case Dungeons:
        {
            return stats->dungeons;
        }
        case None:
        default:
        {
            return 0;
        }
    }
}

void setStatsField(OStats stats, OStatsField field, int value)
{
    switch (field)
    {
        case Tower:
        {
            stats->tower += value;
            break;
        }
        case Wall:
        {
            stats->wall += value;
            
            if (stats->wall < 0)
            {
                stats->tower += stats->wall;
                stats->wall = 0;
            }
            break;
        }
        case Bricks:
        {
            stats->bricks += value;
            
            if (stats->bricks < 0)
            {
                stats->bricks = 0;
            }
            break;
        }
        case Gems:
        {
            stats->gems += value;
            
            if (stats->gems < 0)
            {
                stats->gems = 0;
            }
            break;
        }
        case Recruits:
        {
            stats->recruits += value;
            
            if (stats->recruits < 0)
            {
                stats->recruits = 0;
            }
            break;
        }
        case Quarries:
        {
            stats->quarries += value;
            
            if (stats->quarries < 0)
            {
                stats->quarries = 0;
            }
            break;
        }
        case Magics:
        {
            stats->magics += value;
            
            if (stats->magics < 0)
            {
                stats->magics = 0;
            }
            break;
        }
        case Dungeons:
        {
            stats->dungeons += value;
            
            if (stats->dungeons < 0)
            {
                stats->dungeons = 0;
            }
            break;
        }
        case None:
        default:
        {
            break;
        }
    }
}

void setStats(OStats dest, OStats src)
{
    setStatsField(dest, Tower, src->tower);
    setStatsField(dest, Wall, src->wall);
    setStatsField(dest, Bricks, src->bricks);
    setStatsField(dest, Gems, src->gems);
    setStatsField(dest, Recruits, src->recruits);
    setStatsField(dest, Quarries, src->quarries);
    setStatsField(dest, Magics, src->magics);
    setStatsField(dest, Dungeons, src->dungeons);
}
